### PR TITLE
fixes #527 - Add s3 protocol handler

### LINKED ISSUE #527: Add s3 protocol handler
Eg using this
https://github.com/jenshadlich/S3-URLStreamHandler/blob/master/README.md

### PR DESCRIPTION
**Pull request by AngeloBusato**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -70,6 +70,7 @@ dependencies {
     compileOnly group: 'net.biville.florent', name: 'neo4j-sproc-compiler', version:'1.2'
 
     compile 'com.novell.ldap:jldap:2009-10-07'
+    compile group: 'com.amazonaws', name: 'aws-java-sdk-s3', version: '1.11.241'
```
Comment (jexp) [inline build.gradle]: should this be compileOnly ? so that the user would have to provide the jar?

File: src/main/java/apoc/load/Xml.java [hunk 2]
```diff
@@ -61,8 +63,14 @@
             DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
 
             FileUtils.checkReadAllowed(url);
-            URLConnection urlConnection = new URL(url).openConnection();
-            Document doc = documentBuilder.parse(urlConnection.getInputStream());
+            Document doc;
+            if(url.toLowerCase().startsWith("s3")){
```
Comment (jexp) [inline src/main/java/apoc/load/Xml.java]: don't do any of this here, encapsulate it in file-utils.

File: src/main/java/apoc/util/Util.java [hunk 3]
```diff
@@ -302,6 +307,7 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
     }
 
     public static CountingInputStream openInputStream(String url, Map<String, Object> headers, String payload) throws IOException {
+        if(url.startsWith("s3")) return openS3InputStream(url);
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: can we use a proper URL parsing + getProtocol?

File: src/main/java/apoc/util/s3util/S3Params.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package apoc.util.s3util;
+
+public class S3Params {
+
+    private final String accessKey;
+    private final String secretKey;
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3Params.java]: good question, we should have an option to store this in a separate config, either in neo4j.conf or in a more secure place.

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package apoc.util.s3util;
+
+import com.amazonaws.regions.Regions;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.Objects;
+
+public class S3ParamsExtractor {
+    public static S3Params extract(URL url) throws IOException, IllegalArgumentException {
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: why is this not a method in S3Params?

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        if (!"s3".equals(url.getProtocol())) {
+            throw new IllegalArgumentException("Unsupported protocol '" + url.getProtocol() + "'");
+        }
+
+        //aws credentials
+        String accessKey = null;
+        String secretKey = null;
+
+        if (url.getUserInfo() != null) {
+            String[] credentials = url.getUserInfo().split("[:]");
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: you dont' need the square brackets

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            accessKey = credentials[0];
+            secretKey = credentials[1];
+        } else {
+            if (url.getQuery() != null) {
+                String[] queryStrings = url.getQuery().split("&");
+                if (queryStrings.length == 2) {
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: don't we already have a function for this somewhere? I'm pretty sure.

Comment (jexp) [reply src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: If not a proper version should go into Util or FileUtils

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                                secretKey = valueDecoded;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // bucket
+        String bucket = url.getHost().substring(0, url.getHost().indexOf("."));
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: you can store the host and the index of "." once.

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        // key
+        String key = url.getPath().substring(1);
+
+        // endpoint
+        String endpoint = url.getHost().substring(url.getHost().indexOf(".") + 1);
+
+        String region = null;
+
+        for (Regions r: Regions.values()){
+            if(endpoint.contains(r.getName())){
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: what about case, we should probably use tolower on both. And probably split the endpoint on "." ? and access the right segment.

File: src/main/java/apoc/util/s3util/S3URLConnection.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    public static final String PROP_S3_HANDLER_PROTOCOL = "s3.handler.protocol";
+    public static final String PROP_S3_HANDLER_SIGNER_OVERRIDE = "s3.handler.signerOverride";
+
+    public S3URLConnection(URL url) {
+        super(url);
+    }
+
+    @Override
+    public void connect() throws IOException {
+        // not implemented yet
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3URLConnection.java]: ???

File: src/main/java/apoc/util/s3util/S3URLConnection.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    public static ClientConfiguration buildClientConfig() {
+        final String userAgent = System.getProperty(PROP_S3_HANDLER_USER_AGENT, null);
```
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3URLConnection.java]: are these documented?

File: src/main/java/apoc/util/s3util/S3ParamsExtractor.java (hunk 5 above)
Comment (jexp) [inline src/main/java/apoc/util/s3util/S3ParamsExtractor.java]: is the format documented?

File: src/test/java/apoc/load/LoadJsonTest.java [hunk 12]
```diff
@@ -108,4 +109,34 @@
                     assertFalse(row.hasNext());
                 });
     }
+
+    @Test public void testLoadJsonS3() throws Exception {
+
+        String accessKey = "Q3AM3UQ867SPQQA43P2F";
```
Comment (jexp) [inline src/test/java/apoc/load/LoadJsonTest.java]: Can we unify the minio setup in one test helper class?

**Comment by AngeloBusato** [comment]
Ok @jexp we'll have a look

**Comment by AngeloBusato** [comment]
@jexp  The increased size of apoc jar with aws-sdk-s3 library included is about 4 MBytes.

**Comment by jexp** [comment]
Please have a look at the comments. And we would add the s3 library as an external dependency. Does it has any transitive dependencies?

**Comment by AngeloBusato** [comment]
@jexp PR updated.

with `aws-java-sdk-s3-1.11.250.jar` we have to import these jars too: 

* aws-java-sdk-core-1.11.250.jar 
* httpclient-4.4.8.jar 
* httpcore-4.5.4.jar 
* joda-time-2.9.9.jar

**Comment by jexp** [comment]
Weird:

- joda-time is in the JDK these days (since 1.7)
- httpclient/httpcore would be perhaps ok as a default dependency (not sure how big they are)
- currently I avoided pulling them in and used http-url-connection instead

**Comment by AngeloBusato** [comment]
@jexp The increased size of apoc jar with httpclient/httpcore  libraries included is about 1 MBytes.

**Comment by jexp** [comment]
thanks a lot, merged manually.

**Pull request closed without merging**